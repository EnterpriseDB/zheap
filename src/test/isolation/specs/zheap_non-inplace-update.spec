# Simple tests that show the behavior of snapshots working correctly
# for non-inplace-updates.

setup
{
 CREATE TABLE animals (name text, counter int) WITH (storage_engine = 'zheap');
 CREATE INDEX idx_animals_counter ON animals USING BTREE(counter);
 INSERT INTO animals VALUES ('cat', 1), ('dog', 1), ('monkey', 1);
}

teardown
{
 DROP TABLE animals;
}

session "s1"
setup		{ BEGIN; }
step "r1"	{ SELECT * FROM animals; }
step "c1"	{ COMMIT; }

# index key update
session "s2"
setup       { BEGIN; }
step "w2"	{ UPDATE animals SET counter = counter + 1 WHERE name = 'cat'; }
step "r2"	{ SELECT * FROM animals; }
step "c2"	{ COMMIT; }

# tuple size increase
session "s3"
setup       { BEGIN; }
step "w3"	{ UPDATE animals SET name = 'cat1' WHERE name = 'cat'; }
step "r3"	{ SELECT * FROM animals; }
step "c3"	{ COMMIT; }

# s1 sees previous version of "cat" tuple until the new version is committed
# but s2 sees its own uncommitted data
permutation "r1" "w2" "r2" "r1" "c2" "r1" "c1" "r1"  "c3"

# same thing, but for tuple size increase
permutation "r1" "w3" "r3" "r1" "c3" "r1" "c1" "r1"  "c2"
