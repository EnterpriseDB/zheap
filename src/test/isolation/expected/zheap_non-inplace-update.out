Parsed test spec with 3 sessions

starting permutation: r1 w2 r2 r1 c2 r1 c1 r1 c3
step r1: SELECT * FROM animals;
name           counter        

cat            1              
dog            1              
monkey         1              
step w2: UPDATE animals SET counter = counter + 1 WHERE name = 'cat';
step r2: SELECT * FROM animals;
name           counter        

dog            1              
monkey         1              
cat            2              
step r1: SELECT * FROM animals;
name           counter        

cat            1              
dog            1              
monkey         1              
step c2: COMMIT;
step r1: SELECT * FROM animals;
name           counter        

dog            1              
monkey         1              
cat            2              
step c1: COMMIT;
step r1: SELECT * FROM animals;
name           counter        

dog            1              
monkey         1              
cat            2              
step c3: COMMIT;

starting permutation: r1 w3 r3 r1 c3 r1 c1 r1 c2
step r1: SELECT * FROM animals;
name           counter        

cat            1              
dog            1              
monkey         1              
step w3: UPDATE animals SET name = 'cat1' WHERE name = 'cat';
step r3: SELECT * FROM animals;
name           counter        

dog            1              
monkey         1              
cat1           1              
step r1: SELECT * FROM animals;
name           counter        

cat            1              
dog            1              
monkey         1              
step c3: COMMIT;
step r1: SELECT * FROM animals;
name           counter        

dog            1              
monkey         1              
cat1           1              
step c1: COMMIT;
step r1: SELECT * FROM animals;
name           counter        

dog            1              
monkey         1              
cat1           1              
step c2: COMMIT;
